File: javascript/fibonacci.js
function fibonacci(num) {
  if (num < 2) {
    return num;
  }

  let arr = [0, 1];

  for (let i = 2; i <= num; i++) {
    sum = arr[0] + arr[1];
    arr = [arr[1], sum];
  }

  return arr[1];
}

// Overarching strategy:
//    if the number is 0 or 1, just return the number.
//    use an array to hold placeholder values for 2 numbers
//    sum the numbers, then replace the values so that you're tracking the most recent sum and the previous value
//    keep following this pattern until you reach the desired number
// Steps:
// create an array to hold two values. initially set them to 0 and 1
// iterate through the array num-1 times (or while the iterator is less than num)
// start the iterator at 2 because this is the "starting index" (you've already hard-coded the first 2 numbers in the sequence)
// calculate the sum of the values in arr
// move the second arr value to first space
// put the newNumber in the second space
// increment the iterator
// return the most recent newNumber

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 0");
  console.log("=>", fibonacci(0));

  console.log("");

  console.log("Expecting: 1");
  console.log("=>", fibonacci(2));

  console.log("");

  console.log("Expecting: 55");
  console.log("=>", fibonacci(10));
}

module.exports = fibonacci;

// Please add your pseudocode to this file
// And a written explanation of your solution

File: ruby/fibonacci.rb
require 'pry'

# def fibonacci(num)
#   arr = [0, 1]
#   i = 2
# 
#   while i <= num
#     new_number = arr[i-2] + arr[i-1]
#     arr.push(new_number)
#     i+=1
#   end
# 
#   arr.last
# end 

#   # create an empty array to hold the fibonacci sequence. manually enter the first 2 numbers bc they don't follow the pattern
#   # start the iterator at 2 because that's the index you're starting at
#     # get the next number by summing the previous two numbers
#     # add the new number to the end of the array
#     # increment the iterator
#   # return the last number in the array

def fibonacci(num)
  arr = [0, 1]

  return num if num < 2
  
  (num-1).times do
    sum = arr[0] + arr[1]
    arr = [arr[1], sum]
  end

  arr[1]
end

# create an array to hold two values. initially set them to 0 and 1 because these numbers start the sequence
# loop through the array num-1 times
# calculate the new number using the two existing numbers in the array
# put the value of the second number in place of the first
# put the new number in place of the second
# keep going until you've looped through num-1 times because of zero indexing
# return the most recent "new number" (sum of the previous two)

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 0"
  puts "=>", fibonacci(0)

  puts

  puts "Expecting: 1"
  puts "=>", fibonacci(2)

  puts

  puts "Expecting: 55"
  puts "=>", fibonacci(10)

  # Don't forget to add your own!
end
# Please add your pseudocode to this file
# And a written explanation of your solution
